### PR TITLE
fix region offset handling

### PR DESCRIPTION
**Pull request by jlevon**
The specification states that the region offset given in the region info should
be used as the "offset" when mmap()ing the region from the client side. However,
the library instead implemented a fixed offset scheme  similar to that of vfio -
and no clients actually set up the file like that.

Instead, let servers define their own offsets, and pass them through to clients
as is. It's up to the server to decide how its backing file or files is
organized.

Signed-off-by: John Levon <john.levon@nutanix.com>

### CODE REVIEW

**Comment by jlevon** [comment]
Fixes #478

File: samples/server.c [hunk 1]
```diff
@@ -491,15 +491,29 @@ int main(int argc, char *argv[])
...
+    migr_regs_size = vfu_get_migr_register_area_size();
+    migr_data_size = page_align(bar1_size + sizeof(time_t));
+    migr_size = migr_regs_size + migr_data_size;
+
+    /*
+     * We choose to use a single file which contains both bar1 and the migration
+     * registers. They could also be completely different files.
+     */
+    if (ftruncate(fileno(fp), server_data.bar1_size + migr_size) == -1) {
         err(EXIT_FAILURE, "failed to truncate BAR1 file");
```
Comment (tmakatos) [inline samples/server.c:513]: nit: BAR1

File: samples/server.c [hunk 2]
```diff
@@ -491,15 +491,29 @@ int main(int argc, char *argv[])
...
     }
     server_data.bar1 = mmap(NULL, server_data.bar1_size, PROT_READ | PROT_WRITE,
-                            MAP_SHARED, fileno(bar1_fp), 0);
+                            MAP_SHARED, fileno(fp), 0);
     if (server_data.bar1 == MAP_FAILED) {
         err(EXIT_FAILURE, "failed to mmap BAR1");
```
Comment (tmakatos) [inline samples/server.c:518]: ditto

**Review by tmakatos — APPROVED** [review]

File: samples/server.c (hunk 1 above)
Comment (jlevon) [reply samples/server.c]: Yes, thanks, I re-ordered the comments a bit too.

File: samples/server.c (hunk 2 above)
Comment (jlevon) [reply samples/server.c:518]: this one is fine I think, we are mapping BAR1 from the backing file only.

**Pull request merged**